pf4j/src/main/java/org/pf4j/PluginClassLoader.java

```diff
@@ -255,16 +256,22 @@ public URL getResource(String name) {
 
         return null;
     }
-    
+
     private URL findResourceFromDependencies(String name) {
         log.trace("Search in dependencies for resource '{}'", name);
         List<PluginDependency> dependencies = pluginDescriptor.getDependencies();
         for (PluginDependency dependency : dependencies) {
             PluginClassLoader classLoader = (PluginClassLoader) pluginManager.getPluginClassLoader(dependency.getPluginId());
 
             // If the dependency is marked as optional, its class loader might not be available.
-            if (classLoader == null || dependency.isOptional()) {
-                continue;
+            if (classLoader == null) {
+                if (dependency.isOptional()) {
+                    continue;
+                } else {
+                    throw new PluginRuntimeException(
```

Comment:
This throw of `PluginRuntimeException` could break the hierarchical nature of  `PluginClassLoader` and the current behavior. So, it's a serious modification and we should analyse it very carefully.

Again, a unit test will bring more light.

Now, I don't understand very clear current lines:
```
            // If the dependency is marked as optional, its class loader might not be available.	            // If the dependency is marked as optional, its class loader might not be available.
            if (classLoader == null || dependency.isOptional()) {
```

Maybe @pinhead84 can help us (the lines have been introduced in #270).

Reply:
Hey decebals.

> Now, I don't understand very clear current lines:
>
> ```java
> // If the dependency is marked as optional, its class loader might not be available.
> if (classLoader == null || dependency.isOptional()) {
> ```

If an optional plugin dependency is not present at runtime, the `classLoader` variable is `null`. The method should not throw an exception in this case and just continue its work. 

I'm a bit confused about the logical OR in this IF statement. In my [original commit](https://github.com/pf4j/pf4j/pull/270/commits/9fd2016be4cf8534ff81d13bca65bf2d0d6dafe1) I've used an AND. An OR doesn't really make sense to me - or I don't understand it, yet. Of course, if `classLoader != null && dependency.isOptional() == true`, the method should collect the resources instead of calling `continue`. Otherwise some resources might get ignored. Looks like a bug to me at first sight.

The IF statement has been changed with [this commit](https://github.com/pf4j/pf4j/commit/2f8343cfda32e282a9a75f7e83cd6b05b8aea93a#diff-1e3c97d1c95c106503725b202b9399bdL219).

> This throw of `PluginRuntimeException` could break the hierarchical nature of `PluginClassLoader` and the current behavior. So, it's a serious modification and we should analyse it very carefully.

I'm not a fan of those runtime exceptions. It also looks inconsistent to me as we don't throw any exception if `classLoader.loadClass(className);` fails with a `ClassNotFoundException`. Why not just do some logging here?

Best Regards.

Reply:
> I'm a bit confused about the logical OR in this IF statement. In my [original commit](https://github.com/pf4j/pf4j/pull/270/commits/9fd2016be4cf8534ff81d13bca65bf2d0d6dafe1) I've used an AND. An OR doesn't really make sense to me - or I don't understand it, yet. Of course, if `classLoader != null && dependency.isOptional() == true`, the method should collect the resources instead of calling `continue`. Otherwise some resources might get ignored. Looks like a bug to me at first sight.

BTW: Now I've realized, that @valkuc found the problem introduced with [this commit](https://github.com/pf4j/pf4j/commit/2f8343cfda32e282a9a75f7e83cd6b05b8aea93a#diff-1e3c97d1c95c106503725b202b9399bdL219) and already fixed it - well done! The method now behaves again as it was originally intended (besides the `PluginRuntimeException`).

Reply:
> I'm a bit confused about the logical OR in this IF statement.

The same feeling :smile:.

Reply:
> BTW: Now I've realized, that @valkuc found the problem introduced with this commit and already fixed it - well done! The method now behaves again as it was originally intended (besides the PluginRuntimeException).

I think that:
```java
if (classLoader == null && dependency.isOptional()) {
    continue;
}
```

is enough.

Reply:
If you just replace OR with AND as it was originally then there can be NPE next in code when classloader is null and dependency is not optional. But if this situation simple can't occur in real world, then agree, using just AND will be enough.